src/yunohost/backup.py

```diff
@@ -199,257 +358,506 @@ def _clean_tmp_dir(retcode=0):
-        _clean_tmp_dir(1)
-        raise MoulinetteError(errno.EINVAL, m18n.n('backup_nothings_done'))
-
-    # Calculate total size
-    backup_size = int(subprocess.check_output(
-        ['du', '-sb', tmp_dir]).split()[0].decode('utf-8'))
-    info['size'] = backup_size
-
-    # Create backup info file
-    with open("%s/info.json" % tmp_dir, 'w') as f:
-        f.write(json.dumps(info))
-
-    # Create the archive
-    if not no_compress:
-        logger.info(m18n.n('backup_creating_archive'))
-
-        # Check free space in output directory at first
-        avail_output = subprocess.check_output(
-            ['df', '--block-size=1', '--output=avail', tmp_dir]).split()
+            # Remove added path from csv
+            # TODO
+        else:
+            # Add app info
+            i = app_info(app_instance_name)
+            self.apps_return[app_instance_name] = {
+                'version': i['version'],
+                'name': i['name'],
+                'description': i['description'],
+            }
+        finally:
+            filesystem.rm(tmp_script, force=True)
+            filesystem.rm(env_dict["YNH_BACKUP_CSV"], force=True)
+
+    def _compute_backup_size(self):
+        """ Compute backup size """
+        # FIXME Database dump will be loaded, so dump should use almost the
+        # double of their space
+        # FIXME Some archive will set up dependencies, those are not in this
+        # size info
+        self.size = 0
+        for row in self.paths_to_backup:
+            if row['dest'] != "info.json":
+                self.size += int(subprocess.check_output(['du', '-sb', row['source']])
+                           .split()[0].decode('utf-8'))
+        return self.size
+
+
+class BackupMethod(object):
+    """
+    Abstract class
+    """
+    def __init__(self, repo = None):
+        self.repo = ARCHIVES_PATH if repo is None else repo
+
+    @property
+    def method_name(self):
+        raise MoulinetteError(errno.EINVAL, m18n.n('backup_abstract_method'))
+
+    @property
+    def name(self):
+        return self.manager.name
+
+    @property
+    def work_dir(self):
+        return self.manager.work_dir
+
+    def need_mount(self):
+        return False
+
+    def backup(self, backup_manager):
+        self.manager = backup_manager
+        if self.need_mount():
+            self._mount_csv_listed_files()
+
+    def mount(self, restore_manager):
+        self.manager = restore_manager
+
+    def _check_is_enough_free_space(self):
+        """ Check free space in output directory at first """
+        backup_size = self.manager.size
+        cmd = ['df', '--block-size=1', '--output=avail', self.repo]
+        avail_output = subprocess.check_output(cmd).split()
         if len(avail_output) < 2 or int(avail_output[1]) < backup_size:
             logger.debug('not enough space at %s (free: %s / needed: %d)',
                          output_directory, avail_output[1], backup_size)
```

Comment:
This line will fail if `len(avail_output) < 2`, because then `avail_output[1]` doesn't exists :confused: 

Also it'd be nice for tests, if we could factorize out the part where we fetch the available space (or not use subprocess with df) in a separate function. (So that we can mock the output to be 0 for instance) :D